File: files/CH-159.py
import pandas as pd

path = "CH-159 Data Cleaning.xlsx"
input = pd.read_excel(path, usecols="C:I", skiprows=1, nrows=25)
test = pd.read_excel(path, usecols="K:N", skiprows=1, nrows=4)

input = input.melt(id_vars=["Date"], var_name="Sensor", value_name="Temperature").sort_values(by=["Sensor", "Date"])
input['group'] = input.groupby('Sensor')['Temperature'].diff().ne(0).cumsum()
result = input[input.groupby(['Sensor', 'group'])['Temperature'].transform('size') >= 4].groupby('group').agg(
    From=('Date', 'min'),
    TO=('Date', 'max'),
    Sensor=('Sensor', 'first'),
    Temperature=('Temperature', 'first')
).reset_index(drop=True)

print(result.equals(test)) # True

File: files/100-199/159/CH-159.R
library(tidyverse)
library(readxl)

path = "files/CH-159 Data Cleaning.xlsx"
input = read_excel(path, range = "C2:I27")
test  = read_excel(path, range = "K2:N6")

result = input %>%
  pivot_longer(-Date, names_to = "Sensor", values_to = "Temperature") %>%
  arrange(Sensor, Date) %>%
  mutate(group = cumsum(lag(Temperature, default = first(Temperature)) != Temperature), .by = Sensor) %>%
  mutate(length = n(), .by = c(Sensor, group)) %>%
  filter(length >= 4) %>%
  summarise(From = min(Date), TO = max(Date), Temperature = first(Temperature), Sensor = first(Sensor), .by = group) %>%
  select(From, TO, Sensor, Temperature)

all.equal(result, test)
#> [1] TRUE